src/Controller/IncentiveImportController.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+namespace App\Controller;
+
+use App\Entity\IncentiveImportRow;
+use App\Entity\IncentiveImport;
+use App\Form\IncentiveImportConfirmFormType;
+use App\Form\IncentiveImportFormType;
+use App\Service\IncentiveImportService;
+use App\Service\LoggerService;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\SubmitButton;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Routing\Annotation\Route;
+use App\Audit\Log;
+
+class IncentiveImportController extends BaseController
+{
+    public function __construct(EntityManagerInterface $em)
+    {
+        parent::__construct($em);
+    }
+
+    /**
+     * @Route("/incentive/import", name="incentiveImport", methods={"GET","POST"})
+     */
+    public function incentiveImport(
+        Request $request,
+        SessionInterface $session,
+        LoggerService $loggerService,
+        IncentiveImportService $incentiveImportService
+    ) {
+        $form = $this->createForm(IncentiveImportFormType::class);
+        $form->handleRequest($request);
+        if ($form->isSubmitted()) {
+            $file = $form['incentive_csv']->getData();
+            $fileName = $file->getClientOriginalName();
+            $incentives = [];
+            $incentiveImportService->extractCsvFileData($file, $form, $incentives);
```

Comment:
Instead of passing in an empty array by reference that is modified inside the function, can we just have this method return the resulting array?

Reply:
We are passing both $form & $incentives as reference. $form is passed to get the error messages. That's why I passed both as reference.

Reply:
What I didn't understand is why the need to create the $incentives array outside the method at all, instead of just returning and assigning it?

```php
$incentives = $incentiveImportService->extractCsvFileData($file, $form);
```

and then for the `extractCsvFileData` method:

```php
public function extractCsvFileData($file, $form): void
{
    $incentives = [];
    // ...
    return $incentives;
}
```

Also, unless I'm misunderstanding something, since the form variable is an object, isn't that passed by reference anyways? I don't think you need the `&` in the method signature.